libjwt/jwt-openssl.c

```diff
@@ -364,18 +371,27 @@ int jwt_verify_sha_pem(jwt_t *jwt, const char *head, const char *sig_b64)
 		return EINVAL;
 	}
 
-	sig = jwt_b64_decode(sig_b64, &slen);
-	if (sig == NULL)
-		VERIFY_ERROR(EINVAL);
-
-	bufkey = BIO_new_mem_buf(jwt->key, jwt->key_len);
-	if (bufkey == NULL)
-		VERIFY_ERROR(ENOMEM);
-
-	/* This uses OpenSSL's default passphrase callback if needed. The
-	 * library caller can override this in many ways, all of which are
-	 * outside of the scope of LibJWT and this is documented in jwt.h. */
-	pkey = PEM_read_bio_PUBKEY(bufkey, NULL, NULL, NULL);
+    sig = jwt_b64_decode(sig_b64, &slen);
+    if (sig == NULL)
+        VERIFY_ERROR(EINVAL);
+    
+    if (strncmp(jwt->key, "{", 1) == 0) {
```

Comment:
This is not the way to do this. A random key has a 1-in-256 chance of having "{" as the first character.

Rather than invisibly overloading the current method, what this really needs is an alternative to `jwt_decode()` that accepts a JWKS string instead of key/len params.

Also, this implementation needs to be in `libjwt/jwt.c` so that both the openssl and gnutls variants get this functionality. The only thing that should be in jet-openssl.c is the crypto callbacks, nothing specific to JWT/JWKS.

Reply:
`jwt_decode` is already overloaded for various signing algorithms. It can accept a key to perform an HMAC based verification or a PEM encoded key for RSA and EC based verification. In that context I think it is OK to use it also for passing a JWKS encoded key. Introducing new functions would mean a lot of code duplication.

In my opinion, it should be enough to update the documentation of this function, to explain that a JWKS encoded key is also possible. 

Considering that for RSA verification the key is either a base64 encoded string or a JWKS, I think it is safe to test the first character. The reason is that a base64 encoded string will never contain curly brackets.

If you have any better suggestions, please let me know.

Regarding moving these functions to `jwt.c`. I can move them there, that is not a problem. But since these functions use OpenSSL functions, I thought this was the right place to put them. Correct me if I'm wrong.

Reply:
For some algorithms, the key is a random string, which may start with '{' and jwt_decode() is not "overloaded" it knows which key you are passing it because you tell it the algorithm to use. OpenSSL and GnuTLS internally handle the types.